### PR TITLE
Add Mistral model support

### PR DESCRIPTION
**Pull request by yqf66**
#### Before submitting

- [ ] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [x] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

## PR Type

New model support.

## Description

This PR adds Mistral model support to PaddleFormers, including:

- `MistralConfig`
- `MistralModel`
- `MistralForCausalLM`
- `MistralForSequenceClassification`
- `MistralForTokenClassification`
- `MistralForQuestionAnswering`
- Pipeline-compatible `MistralForCausalLMPipe`
- AutoConfig / AutoModel registration
- Mistral LoRA target module patterns
- Unit tests and CI model option
- Integration utilities for slim-model training validation

### 1. Unit Tests

Command:

```bash
PYTHONPATH=/sda/yuqifan/PaddleFormer/PaddleFormers-pr \
python -m unittest tests.transformers.mistral.test_modeling
```

Result:

```text
Ran 6 tests in 0.836s
OK
```

### 2. Static Checks

Commands:

```bash
python -m py_compile \
  paddleformers/transformers/mistral/__init__.py \
  paddleformers/transformers/mistral/configuration.py \
  paddleformers/transformers/mistral/modeling.py \
  tests/transformers/mistral/test_modeling.py \
  tests/integration_test/prepare_mistral_slim_model.py

git diff --check
```

Result: passed.

### 3. Same-Token 300-Step Training A/B

To isolate model migration behavior from template and dataloader differences, a minimal same-token A/B validation was run with:

- Same slim Mistral model: 2 decoder layers
- Same fixed `input_ids` / `labels`
- AdamW
- Learning rate: `1e-5`
- Warmup steps: `20`
- Gradient accumulation steps: `4`
- Total optimizer steps: `300`
- Precision path: fp32 parameters + bf16 autocast
- Fused/flash fast paths disabled for stronger operator alignment

Loss comparison uses:

```text
diff = paddle_loss - torch_loss
```

Results:

| Range | n | mean_diff | std_diff | max_abs_diff |
| --- | ---: | ---: | ---: | ---: |
| step >= 1 | 300 | -0.002846 | 0.040334 | 0.285791 |
| step >= 21 | 280 | +0.001544 | 0.005616 | 0.019010 |
| step >= 51 | 250 | +0.002272 | 0.005101 | 0.019010 |
| step >= 101 | 200 | +0.000105 | 0.000203 | 0.001271 |

Conclusion:

- Early steps still contain transient numerical differences.
- After warmup/early transient, Paddle and Torch training loss curves are highly aligned.
- From step 101 onward, the difference is around `1e-4 ~ 1e-3`, indicating stable migration behavior.

Detailed validation notes are included in:

```text
tests/integration_test/mistral_same_token_ab_300_report.md
```

## Checklist

- [x] Add model implementation
- [x] Add auto mapping
- [x] Add unit tests
- [x] Add CI model option
- [x] Add slim-model preparation script
- [x] Add 300-step SFT validation config/script
- [x] Run local unit tests
- [x] Run static checks

### CODE REVIEW

**Comment by yqf66** [comment]
提交：be81eb1 Fix Mistral lint issues

另外两个失败的 GPU CI 日志显示是 CondaHTTPError 403，发生在创建 conda 环境阶段，尚未进入 Mistral 测试逻辑，属于 CI 环境/网络问题，可请维护者 rerun。

File: paddleformers/transformers/mistral/modeling.py [hunk 1]
```diff
@@ -0,0 +1,885 @@
...
+            output_hidden_states=output_hidden_states,
+            return_dict=True,
+        )
+
+        hidden_states = outputs[0]
+        logits = self.lm_head(hidden_states)
+
+        loss = None
+        if labels is not None:
+            loss, _ = self.criterion(logits, labels)
```
Comment (risemeup1111) [inline paddleformers/transformers/mistral/modeling.py]: P1 **优先级：P1**
这里声明了 `loss_mask` 参数，但计算 loss 时没有传给 `CriterionLayer`。Trainer 会把 batch dict 以 `model(**inputs)` 形式传入，管线并行路径也把 `loss_mask` 作为末 stage 输入键；当前实现会退化为只按 `labels != -100` 生成 mask，导致依赖显式 `loss_mask` 的 SFT/预训练样本把本应屏蔽的 token 计入损失。请把 `loss_mask` 继续传入 criterion，并补一个带 `loss_mask` 的 CausalLM loss 单测。

```suggestion
            loss, _ = self.criterion(logits, labels, loss_mask)
```

File: paddleformers/transformers/mistral/modeling.py [hunk 2]
```diff
@@ -0,0 +1,885 @@
...
+        if output_hidden_states:
+            all_hidden_states.append(hidden_states)
+
+        all_hidden_states = tuple(all_hidden_states) if all_hidden_states else None
+
+        if not return_dict:
+            outputs = [hidden_states]
+            if output_hidden_states:
+                outputs.append(all_hidden_states)
+            return tuple(outputs)
```
Comment (risemeup1111) [inline paddleformers/transformers/mistral/modeling.py]: P2 **优先级：P2**
`MistralModel` 在 `return_dict=True` 时会通过 `BaseModelOutputWithPast.past_key_values` 返回 cache，但 tuple 路径只返回了 `hidden_states` 和可选 `hidden_states` 历史；因此直接调用 `MistralModel(..., use_cache=True, return_dict=False)` 的用户拿不到本轮填充后的 `past_key_values`。这和 `BaseModelOutputWithPast` 的 tuple 语义不一致，也会让依赖 tuple 输出做增量解码的调用方失效。建议像 Qwen2/Phi3 的 base model 一样把 `past_key_values` 放进非空输出序列。

```suggestion
            return tuple(v for v in [hidden_states, past_key_values, all_hidden_states] if v is not None)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已完成本轮审查，发现一个会影响训练 loss 口径的阻塞问题，需要先修复后再合入。另有一个非阻塞的 base model cache tuple 输出兼容性问题，细节都已放在行级评论里。

本地环境缺少 `paddle`，因此未能在当前 worktree 运行新增单测；CI 当前显示通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by yqf66** [comment]
1. CausalLM loss 计算时已将 loss_mask 继续传入 CriterionLayer，避免 SFT/预训练样本中显式 loss_mask 被忽略。
2. MistralModel 在 return_dict=False 时的 tuple 输出已补充 past_key_values，与 BaseModelOutputWithPast 语义保持一致。
3. 补充了 loss_mask 传递和 tuple cache 输出的回归单测。

**Review by risemeup1111 — APPROVED** [review]
已复查新提交。上轮提出的 `loss_mask` 传递和 base model cache tuple 输出问题都已在当前代码中修复，并补充了对应回归测试；未发现新的需要阻塞合入的问题。

我在当前环境运行了相关文件的 `py_compile`，结果通过；由于本地仍缺少 `paddle`，未能实际运行 `tests.transformers.mistral.test_modeling`。CI 当前显示通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
